kubeadm: add 1.32 documentation (#48408)
kubernetes/website: content/en/docs/reference/setup-tools/kubeadm/implementation-details.md

  - `front-proxy-ca.crt`, `front-proxy-ca.key` for the front proxy certificate authority
  - `front-proxy-client.crt`, `front-proxy-client.key` for the front proxy client

> ## The kubeadm configuration API

Reviewer: ```suggestion
## The kubeadm configuration file format
```
We stretch the term “application programming interface“ really far. These are really configuration files that happen to align with our API conventions.

Most kubeadm commands support a `--config` flag which allows passing a configuration file from
disk. The configuration file format follows the common Kubernetes API `apiVersion` / `kind` scheme,

> but is considered a component configuration API, similarly to that of the kubelet.

Reviewer: ```suggestion
but is considered a component configuration format. Several Kubernetes components, such as the kubelet,
also support file-based configuration.
```

> Different kubeadm commands require a different `kind`. For example, `InitConfiguration` for `kubeadm init`, `JoinConfiguration` for `kubeadm join`, `UpgradeConfiguration` for `kubeadm upgrade` and `ResetConfiguration`

Reviewer: (nit)
```suggestion
Different kubeadm subcommands require a different `kind` of configuration file.
For example, `InitConfiguration` for `kubeadm init`, `JoinConfiguration` for `kubeadm join`, `UpgradeConfiguration` for `kubeadm upgrade` and `ResetConfiguration`
```

for `kubeadm reset`.

> The command `kubeadm config migrate` can be used to migrate an older, deprecated API file
> to a newer API file.

Reviewer: Is this right?
```suggestion
The command `kubeadm config migrate` can be used to migrate an older format configuration
file to a newer (current) configuration format. The kubeadm tool only supports migrating from
deprecated configuration formats to the current format.
```
I reworded into the active voice per the [style guide](https://kubernetes.io/docs/contribute/style/style-guide/).

> See the [API reference](/docs/reference/config-api/kubeadm-config.v1beta4/) page for more details.

Reviewer: ```suggestion
See the [kubeadm configuration reference](/docs/reference/config-api/kubeadm-config.v1beta4/) page for more details.
```

See [RBAC user facing role bindings](/docs/reference/access-authn-authz/rbac/#user-facing-roles)
for additional information on RBAC and built-in ClusterRoles and groups.

> To generate kubeconfig files for additional users the command `kubeadm kubeconfig user` can be used.

Reviewer: Does this look right?
```suggestion
You can run [`kubeadm kubeconfig user`](/docs/reference/setup-tools/kubeadm/kubeadm-kubeconfig/#cmd-kubeconfig-user)
to generate kubeconfig files for additional users.

{{< caution >}}
The generated configuration files include an embedded authentication key, and you should treat
them as confidential.
{{< /caution >}}
```

> Please note that:

Reviewer: Optional extra tweak:
```suggestion
Also note that:
```

- Leader election is enabled for both the controller-manager and the scheduler
- Controller-manager and the scheduler will reference kubeconfig files with their respective, unique identities

> - All static Pods get any extra flags or patches specified by the user as described in

Reviewer: (nit / optional style improvement)
```suggestion
- All static Pods get any extra flags or patches that you specify, as described in
```

> Please note that:

Reviewer: Optional extra rewording:
```suggestion
Also note that:
```

    compared with the CA retrieved initially

{{< note >}}

> CA validation can be skipped by passing `--discovery-token-unsafe-skip-ca-verification` flag;

Reviewer: ```suggestion
You can skip CA validation by passing the `--discovery-token-unsafe-skip-ca-verification` flag on the command line.
```

- The automatic CSR approval is managed by the csrapprover controller, according to
  the configuration present in the `kubeadm init` process
{{< /note >}}

## kubeadm upgrade workflow internal design

`kubeadm upgrade` has sub-commands for handling the upgrade of the Kubernets cluster created by kubeadm.

> `kubeadm upgrade apply` must be called on a control plane node, chosen by the user, to start
> the upgrade process and `kubeadm upgrade node` must be called on any remaining control plane or worker nodes.

Reviewer: ```suggestion
Tou must run `kubeadm upgrade apply` on a control plane node (you can choose which one);
this starts the upgrade process. You then run run `kubeadm upgrade node` on all remaining
nodes (both worker nodes and control plane nodes).
```

> Both `kubeadm upgrade apply` and `kubeadm upgrade node` have a `phase` subcommand which grants access

Reviewer: (nit)
```suggestion
Both `kubeadm upgrade apply` and `kubeadm upgrade node` have a `phase` subcommand which provides access
```

to the internal phases of the upgrade process.
See [`kubeadm upgrade phase`](/docs/reference/setup-tools/kubeadm/kubeadm-upgrade-phase/) for more details.

Additional utility upgrade commands are `kubeadm upgrade plan` and `kubeadm upgrade diff`.

All upgrade sub-commands support passing a configuration file.

### kubeadm upgrade plan

> Can be optionally called before `kubeadm upgrade apply`. It checks which versions are available to upgrade

Reviewer: ```suggestion
You can optionally run `kubeadm upgrade plan` before you run `kubeadm upgrade apply`.
The `plan` subcommand checks which versions are available to upgrade
```

to and validates whether your current cluster is upgradeable.

### kubeadm upgrade diff

Shows what differences would be applied to existing static pod manifests for control plane nodes.

> A more verbose option to do the same is running `kubeadm upgrade apply --dry-run` or

Reviewer: ```suggestion
A more verbose way to do the same thing is running `kubeadm upgrade apply --dry-run` or
```

- Runs preflight checks similarly to `kubeadm init` and `kubeadm join` ensuring container images are downloaded
and the cluster is in a good state to be upgraded.
- Upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests` and waits
for the kubelet to restart the components if the files have changed.
- Uploads updated kubeadm and kubelet configurations to the cluster in the respective `kube-system/kubeadm-config`
and `kube-system/kubelet-config` ConfigMaps.
- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
- Configures bootstrap token and `kube-public/cluster-info` ConfigMap RBAC rules. This is the same as
in the `kubeadm init` stage and ensures that the cluster continues to support nodes joining with bootstrap tokens.
- Upgrades the kube-proxy and CoreDNS addons conditionally if all existing kube-apiservers in the cluster
have already been upgraded to a the target version.
- Performs any post upgrade tasks, such as, release specific cleanups of deprecated features.

### kubeadm upgrade node

`kubeadm upgrade node` upgrades a single control plane or worker node after the cluster upgrade has

> starteed with `kubeadm upgrade apply`. The command detects if the node is a control plane node by checking

Reviewer: ```suggestion
started (by running `kubeadm upgrade apply`). The command detects if the node is a control plane node by checking
```

> if the file `/etc/kubernetes/manifests/kube-apiserver.yaml` exists. That means there is a running kube-apiserver
> on the node.

Reviewer: ```suggestion
if the file `/etc/kubernetes/manifests/kube-apiserver.yaml` exists. On finding that file, the kubeadm tool
infers that there is a running kube-apiserver Pod on this node.
```
?

- Runs preflight checks similarly to `kubeadm upgrade apply`.
- For control plane nodes, upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests`
and waits for the kubelet to restart the components if the files have changed.
- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.

> - For control plane nodes, Upgrades the kube-proxy and CoreDNS addons conditionally if all existing

Reviewer: ```suggestion
- (For control plane nodes) upgrades the kube-proxy and CoreDNS
  {{< glossary_tooltip text="addons" term_id="addons" >}} conditionally, provided that all existing
```

kube-apiservers in the cluster have already been upgraded to a the target version.
- Performs any post upgrade tasks, such as, release specific cleanups of deprecated features.

## kubeadm reset workflow internal design

> The command `kubeadm reset` can be used to reset a node where kubeadm commands previously executed.

Reviewer: ```suggestion
You can use the `kubeadm reset` subcommand a node where kubeadm commands previously executed.
```

> Please note that `kubeadm reset` is considered a best-effort cleanup of the node.

Reviewer: ```suggestion
This subcommand performs a **best-effort** cleanup of the node.
```

If certain actions fail you must intervene and perform manual cleanup.

The command supports phases.
See [`kubeadm reset phase`](/docs/reference/setup-tools/kubeadm/kubeadm-reset-phase/) for more details.

The command supports a configuration file.

Additionally:

> - IPVS and iptables rules are not cleaned.
> - CNI configuration is not cleaned.
> - `.kube/` in the home user directory is not cleaned.

Reviewer: ```suggestion
- IPVS, iptables and nftables rules are **not** cleaned up.
- CNI (network plugin) configuration is **not** cleaned up.
- `.kube/` in the user's home directory is **not** cleaned up.
```

- The automatic CSR approval is managed by the csrapprover controller, according to
  the configuration present in the `kubeadm init` process
{{< /note >}}

## kubeadm upgrade workflow internal design

`kubeadm upgrade` has sub-commands for handling the upgrade of the Kubernets cluster created by kubeadm.

> Tou must run `kubeadm upgrade apply` on a control plane node (you can choose which one);

Reviewer 2: ```suggestion
You must run `kubeadm upgrade apply` on a control plane node (you can choose which one);
```

> this starts the upgrade process. You then run run `kubeadm upgrade node` on all remaining

Reviewer 3: ```suggestion
this starts the upgrade process. You then run `kubeadm upgrade node` on all remaining
```

nodes (both worker nodes and control plane nodes).

Both `kubeadm upgrade apply` and `kubeadm upgrade node` have a `phase` subcommand which provides access
to the internal phases of the upgrade process.
See [`kubeadm upgrade phase`](/docs/reference/setup-tools/kubeadm/kubeadm-upgrade-phase/) for more details.

Additional utility upgrade commands are `kubeadm upgrade plan` and `kubeadm upgrade diff`.

All upgrade sub-commands support passing a configuration file.

### kubeadm upgrade plan

You can optionally run `kubeadm upgrade plan` before you run `kubeadm upgrade apply`.
The `plan` subcommand checks which versions are available to upgrade
to and validates whether your current cluster is upgradeable.

### kubeadm upgrade diff

> Shows what differences would be applied to existing static pod manifests for control plane nodes.

Reviewer 2: ```suggestion
This shows what differences would be applied to existing static pod manifests for control plane nodes.
```

A more verbose way to do the same thing is running `kubeadm upgrade apply --dry-run` or
`kubeadm upgrade node --dry-run`.

### kubeadm upgrade apply

> `kubeadm upgrade apply` takes cane of preparing the cluster for the upgrade of all nodes, and also

Reviewer 2: ```suggestion
`kubeadm upgrade apply` prepares the cluster for the upgrade of all nodes, and also
```

> upgrades the control plane node where its called. The steps it performs are:

Reviewer 2: ```suggestion
upgrades the control plane node where it's run. The steps it performs are:
```

> - Runs preflight checks similarly to `kubeadm init` and `kubeadm join` ensuring container images are downloaded
> and the cluster is in a good state to be upgraded.

Reviewer 2: ```suggestion
- Runs preflight checks similarly to `kubeadm init` and `kubeadm join`, ensuring container images are downloaded
  and the cluster is in a good state to be upgraded.
```

- Upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests` and waits

> for the kubelet to restart the components if the files have changed.

Reviewer 2: ```suggestion
  for the kubelet to restart the components if the files have changed.
```

> - Uploads updated kubeadm and kubelet configurations to the cluster in the respective `kube-system/kubeadm-config`

Reviewer 2: ```suggestion
- Uploads the updated kubeadm and kubelet configurations to the cluster in the `kubeadm-config`
```

> and `kube-system/kubelet-config` ConfigMaps.

Reviewer 2: ```suggestion
  and the `kubelet-config` ConfigMaps (both in the `kube-system` namespace).
```

- Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.

> - Configures bootstrap token and `kube-public/cluster-info` ConfigMap RBAC rules. This is the same as

Reviewer 2: ```suggestion
- Configures bootstrap token and the `cluster-info` ConfigMap for RBAC rules. This is the same as
```

> in the `kubeadm init` stage and ensures that the cluster continues to support nodes joining with bootstrap tokens.

Reviewer 2: ```suggestion
  in the `kubeadm init` stage and ensures that the cluster continues to support nodes joining with bootstrap tokens.
```

- Upgrades the kube-proxy and CoreDNS addons conditionally if all existing kube-apiservers in the cluster

> have already been upgraded to a the target version.

Reviewer 2: ```suggestion
  have already been upgraded to the target version.
```

> - Performs any post upgrade tasks, such as, release specific cleanups of deprecated features.

Reviewer 2: ```suggestion
- Performs any post-upgrade tasks, such as, cleaning up deprecated features which are release specific.
```

### kubeadm upgrade node

`kubeadm upgrade node` upgrades a single control plane or worker node after the cluster upgrade has
started (by running `kubeadm upgrade apply`). The command detects if the node is a control plane node by checking
if the file `/etc/kubernetes/manifests/kube-apiserver.yaml` exists. On finding that file, the kubeadm tool
infers that there is a running kube-apiserver Pod on this node.

- Runs preflight checks similarly to `kubeadm upgrade apply`.
- For control plane nodes, upgrades the control plane manifest files on disk in `/etc/kubernetes/manifests`

> and waits for the kubelet to restart the components if the files have changed.

Reviewer 2: ```suggestion
  and waits for the kubelet to restart the components if the files have changed.
```

> - Writes updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.

Reviewer 2: ```suggestion
- Writes the updated kubelet configuration for this node in `/var/lib/kubelet/config.yaml`.
```

- (For control plane nodes) upgrades the kube-proxy and CoreDNS
  {{< glossary_tooltip text="addons" term_id="addons" >}} conditionally, provided that all existing

> kube-apiservers in the cluster have already been upgraded to a the target version.

Reviewer 2: ```suggestion
  API servers in the cluster have already been upgraded to the target version.
```

> - Performs any post upgrade tasks, such as, release specific cleanups of deprecated features.

Reviewer 2: ```suggestion
- Performs any post-upgrade tasks, such as cleaning up deprecated features which are release specific.
```

## kubeadm reset workflow internal design

> You can use the `kubeadm reset` subcommand a node where kubeadm commands previously executed.

Reviewer 2: ```suggestion
You can use the `kubeadm reset` subcommand on a node where kubeadm commands previously executed.
```
